clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -647,27 +647,27 @@ private long batchReady(boolean exhausted, TopicPartition part, Node leader,
     }
 
     /**
-     * Iterate over partitions to see which one have batches ready and collect leaders of those partitions
-     * into the set of ready nodes.  If partition has no leader, add the topic to the set of topics with
-     * no leader.  This function also calculates stats for adaptive partitioning.
+     * Iterate over partitions to see which one have batches ready and collect leaders of those
+     * partitions into the set of ready nodes.  If partition has no leader, add the topic to the set
+     * of topics with no leader.  This function also calculates stats for adaptive partitioning.
      *
-     * @param metadata The cluster metadata
-     * @param nowMs The current time
-     * @param topic The topic
-     * @param topicInfo The topic info
+     * @param cluster               The cluster metadata
+     * @param nowMs                 The current time
+     * @param topic                 The topic
+     * @param topicInfo             The topic info
      * @param nextReadyCheckDelayMs The delay for next check
-     * @param readyNodes The set of ready nodes (to be filled in)
-     * @param unknownLeaderTopics The set of topics with no leader (to be filled in)
+     * @param readyNodes            The set of ready nodes (to be filled in)
+     * @param unknownLeaderTopics   The set of topics with no leader (to be filled in)
      * @return The delay for next check
      */
-    private long partitionReady(Metadata metadata, long nowMs, String topic,
+    private long partitionReady(Cluster cluster, long nowMs, String topic,
```

Comment:
In some ways, this is a step backwards. We have been trying to reduce the reliance on `Cluster` internally because it is public. With a lot of internal usage, we end up making changes to the API which are only needed for the internal implementation (as we are doing in this PR). Have you considered alternatives? Perhaps we could expose something like `Cluster`, but with a reduced scope?

Reply:
@hachikuji 

Thanks for pointing it out. As it turns out I don't need to extend the public interface of `Cluster` in order to get epoch. So internal usage doesn't change Cluster's interface anymore.

> We have been trying to reduce the reliance on Cluster internally because it is public.

This could be achieved by created a forwarding class `ClusterInternal` that uses `Cluster` by composition offering the same interface. Then `client` code can be refactored to use `ClusterInternal`. 

But this is going to be a size-able PR, how about keeping it separate from this PR? As the intention of this PR is to fix the perf bug.

Reply:
Makes sense. We'd probably have to do it the other way around though I guess? The client's dependence on `Cluster` cannot be easily changed, but we can move the internal implementation anywhere we want.

Reply:
I was looking into your idea a little bit. There might be a simple enough variation that wouldn't require significant changes. What do you think about this? https://github.com/apache/kafka/compare/trunk...hachikuji:kafka:internal-cluster-view?expand=1

Reply:
@hachikuji Thanks for the draft PR. I have introduced `InternalCluster` as a wrapper around public `Cluster`. I have extended `InternalCluster` to `leaderEpochFor` that is only for client's internal-usage.